### PR TITLE
feat: add proxy scanner token management

### PR DESCRIPTION
**Pull request by credibleforce**
Signed-off-by: Lacework <tech-ally@lacework.net>

## Summary

Add support for managing proxy scanner tokens from the API and CLI. This will be used to support new terraform resource for inline and proxy scanner.

## How did you test this change?

API testing using `api/_examples/container-registries/inline_scanner/main.go` and `api/_examples/container-registries/proxy_scanner/main.go`

## Issue

https://lacework.atlassian.net/jira/software/projects/ALLY/issues/ALLY-1037

### CODE REVIEW

File: api/container_registries_proxy_scanner.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	err error,
+) {
+	err = svc.update(data.ID(), data, &response)
+	return
+}
+
+type ProxyScannerIntegrationResponse struct {
+	Data        ProxyScannerIntegration  `json:"data"`
+	ServerToken *ProxyScannerServerToken `json:"serverToken"`
+}
```
Comment (afiune) [inline api/container_registries_proxy_scanner.go:43]: The proxy scanner response also have a `props` field, I wonder if we need it? 🤔

File: api/container_registries_proxy_scanner.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+type ProxyScannerIntegration struct {
+	v2CommonIntegrationData
+	Data        ProxyScannerData         `json:"data"`
+	ServerToken *ProxyScannerServerToken `json:"serverToken"`
+}
+
+type ProxyScannerServerToken struct {
```
Comment (afiune) [inline api/container_registries_proxy_scanner.go:51]: I wonder if we can combine these three structs (inline, proxy, v2) and just have the one you define in `V2ServerToken` instead.

File: cli/cmd/cloud_account.go [hunk 3]
```diff
@@ -187,7 +188,15 @@ func buildDetailsTable(integration api.V2RawType) string {
...
 				for _, i := range val {
-					values = append(values, i.(string))
+					if _, ok := i.(string); ok {
+						values = append(values, i.(string))
+					} else if _, ok := i.(map[string]interface{}); ok {
+						for m, n := range i.(map[string]interface{}) {
+							values = append(values, fmt.Sprintf("%s:%s", m, n))
+						}
+					} else {
+						cli.OutputHuman("[WARN] Unknown type %T for key %s\n", i, k)
```
Comment (afiune) [inline cli/cmd/cloud_account.go:198]:
```suggestion
						cli.Log.Warn("unable to build table details, unknown type", "type", i, "key", k)
```

File: cli/cmd/prompt.go [hunk 4]
```diff
@@ -33,15 +31,15 @@ func promptRequiredStringLen(size int, err string) func(interface{}) error {
...
+// 		i, ok := input.(string)
+// 		if !ok {
+// 			return errors.New(err)
+// 		}
+// 		if _, e := strconv.Atoi(i); e != nil {
+// 			return errors.New(err)
+// 		}
+// 		return nil
+// 	}
+// }
```
Comment (afiune) [inline cli/cmd/prompt.go:45]: feel free to remove this if we don't need it anymore

**Review by afiune — CHANGES_REQUESTED** [review]

File: api/container_registries.go [hunk 5]
```diff
@@ -183,7 +187,12 @@ func (svc *ContainerRegistriesService) Get(guid string, response interface{}) er
 
 type ContainerRegistryRaw struct {
 	v2CommonIntegrationData
-	Data interface{} `json:"data,omitempty"`
+	Data        interface{}   `json:"data,omitempty"`
+	ServerToken V2ServerToken `json:"serverToken,omitempty"`
+}
+type V2ServerToken struct {
```
[the remark this answers is not in the document]
Comment (credibleforce) [reply api/container_registries.go:193]: agreed that this is best defined at the container_registries level and the reused. I updated to fix this.

File: api/container_registries.go [hunk 6]
```diff
@@ -183,7 +187,12 @@ func (svc *ContainerRegistriesService) Get(guid string, response interface{}) er
 
 type ContainerRegistryRaw struct {
 	v2CommonIntegrationData
-	Data interface{} `json:"data,omitempty"`
+	Data        interface{}   `json:"data,omitempty"`
+	ServerToken V2ServerToken `json:"serverToken,omitempty"`
```
[the remark this answers is not in the document]
Comment (credibleforce) [reply api/container_registries.go]: yep, I missed that one. updated to pointer.

File: api/container_registries_inline_scanner.go [hunk 7]
```diff
@@ -38,18 +38,19 @@ func (svc *ContainerRegistriesService) UpdateInlineScanner(data ContainerRegistr
 }
 
 type InlineScannerIntegrationResponse struct {
```
[the remark this answers is not in the document]
Comment (credibleforce) [reply api/container_registries_inline_scanner.go]: Add response processing for this response in get requests and included tests here:
https://github.com/lacework/go-sdk/blob/da3989655d037469a2fc922d797ac1a3b8ef984d/api/_examples/container-registries/inline_scanner/props.go

and here for proxy:
https://github.com/lacework/go-sdk/blob/da3989655d037469a2fc922d797ac1a3b8ef984d/api/_examples/container-registries/proxy_scanner/props.go

File: api/container_registries_proxy_scanner.go (hunk 1 above)
Comment (credibleforce) [reply api/container_registries_proxy_scanner.go:42]: props handling added to inline and proxy scanner to allow visibility into policy assignment. good catch!

File: api/container_registries_proxy_scanner.go (hunk 2 above)
Comment (credibleforce) [reply api/container_registries_proxy_scanner.go]: done

File: cli/cmd/cloud_account.go [hunk 8]
```diff
@@ -182,18 +184,73 @@ func buildDetailsTable(integration api.V2RawType) string {
...
+					} else {
+						cli.Log.Warn("unable to build table details, unknown type", "type", i, "key", k)
+					}
 				}
 				details = append(details, []string{strings.ToUpper(format.SpaceUpperCase(k)), strings.Join(values, ",")})
 			}
 		}
 	}
 
+	if caMap, ok := integration.GetProps().(map[string]interface{}); ok {
```
Comment (afiune) [inline cli/cmd/cloud_account.go:209]: This look good. Thanks for adding the props field.

I think the only missing part is the Server Token that is only available to inline scanner and proxy scanners, so it should not be added to the `V2RawType` struct but instead, identify if the provider interface `integration` is of type `ContainerRegistryRaw` and if so, cast it and access the server token information.

**Review by afiune — CHANGES_REQUESTED** [review]
one more thing and then we are good to go!

tenor-145007785

File: cli/cmd/cloud_account.go (hunk 8 above)
Comment (credibleforce) [reply cli/cmd/cloud_account.go:209]: I added a handler for the `ContainerRegistryRaw` type and output the `SERVER_TOKEN` and `SERVER_TOKE_URI`:
https://github.com/lacework/go-sdk/blob/75c5c324238291cbefe1694de75eb61cbbbb294e/cli/cmd/cloud_account.go#L254-L258

Do we need the URI, I don't know but might as well expose it here while we're at it.

File: cli/cmd/cloud_account.go [hunk 9]
```diff
@@ -182,18 +184,34 @@ func buildDetailsTable(integration api.V2RawType) string {
...
+					}
 				}
 				details = append(details, []string{strings.ToUpper(format.SpaceUpperCase(k)), strings.Join(values, ",")})
 			}
 		}
 	}
 
+	// get server token for container registry type only
+	if c, ok := integration.(api.ContainerRegistryRaw); ok {
+		details = append(details, []string{"SERVER_TOKEN", c.ServerToken.ServerToken})
```
Comment (afiune) [inline cli/cmd/cloud_account.go:211]: This code panic's in CI:
```
                Error:          Should be empty, but was panic: runtime error: invalid memory address or nil pointer dereference                    
                                [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xcaa29f]                                              
                                                                                                                                                    
                                goroutine 1 [running]:                                                                                              
                                github.com/lacework/go-sdk/cli/cmd.buildDetailsTable({0x1283738, 0xc000558100})                                     
                                        /codefresh/volume/go-sdk/cli/cmd/cloud_account.go:211 +0x15f                                                
                                github.com/lacework/go-sdk/cli/cmd.containerRegistryShow(0x18a58a0?, {0xc000466af0, 0x1, 0x1?})                     
                                        /codefresh/volume/go-sdk/cli/cmd/container_registry.go:127 +0x4b7                                           
                                github.com/spf13/cobra.(*Command).execute(0x18a58a0, {0xc000466ad0, 0x1, 0x1})                                      
                                        /codefresh/volume/go-sdk/vendor/github.com/spf13/cobra/command.go:856 +0x67c                                
                                github.com/spf13/cobra.(*Command).ExecuteC(0x18acba0)                                                               
                                        /codefresh/volume/go-sdk/vendor/github.com/spf13/cobra/command.go:974 +0x3b4                                
                                github.com/spf13/cobra.(*Command).Execute(...)                                                                      
                                        /codefresh/volume/go-sdk/vendor/github.com/spf13/cobra/command.go:902                                       
                                github.com/lacework/go-sdk/cli/cmd.Execute()                                                                        
                                        /codefresh/volume/go-sdk/cli/cmd/root.go:159 +0x111                                                         
                                main.main()                                                                                                         
                                        /codefresh/volume/go-sdk/cli/main.go:29 +0x19                                                               

```

**Review by afiune — CHANGES_REQUESTED** [review]

File: cli/cmd/cloud_account.go (hunk 9 above)
Comment (credibleforce) [reply cli/cmd/cloud_account.go:211]: ah got it nil check on the ServerToken for all of the non inline/proxy scanner types.

File: api/_examples/container-registries/inline_scanner/main.go [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+	// 	inlineScannerCreateResponse.Data.IntgGuid,
+	// 	api.MapPolicyRequest{
+	// 		Evaluate:    true,
+	// 		PolicyGuids: []string{"VULN_0595430C23E5C3BBB5EBDB59CEF17467AF592C825562090FDA9"},
+	// 	},
+	// )
+	// if errMap != nil {
+	// 	log.Fatal(errMap)
+	// }
+	// fmt.Printf("Inline scanner map policy guid: %s\n", inlineScannerMapPolicyResponse.Data.IntgGuid)
```
[the remark this answers is not in the document]
Comment (credibleforce) [reply api/_examples/container-registries/inline_scanner/main.go]: removed this

**Pull request merged**